subsys/logging/log_rpc.c

```diff
@@ -0,0 +1,225 @@
+/*
+ * Copyright (c) 2024 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+/**
+ * @file
+ *   This file implements Remote Procedure Calls (RPC) for logging subsystem.
+ *
+ */
+
+#include <nrf_rpc_cbor.h>
+#include <nrf_rpc/nrf_rpc_uart.h>
+
+#include <zephyr/device.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/logging/log_backend.h>
+#include <zephyr/logging/log_backend_std.h>
+#include <zephyr/logging/log_output.h>
+
+#include <string.h>
+
+#ifdef CONFIG_NRF_RPC_IPC_SERVICE
+NRF_RPC_IPC_TRANSPORT(log_rpc_tr, DEVICE_DT_GET(DT_NODELABEL(ipc0)), "log_rpc_ept");
+#elif defined(CONFIG_NRF_RPC_UART_TRANSPORT)
+#define log_rpc_tr NRF_RPC_UART_TRANSPORT(DT_NODELABEL(uart1))
+#endif
+NRF_RPC_GROUP_DEFINE(log_rpc_group, "log", &log_rpc_tr, NULL, NULL, NULL);
+
+enum log_rpc_evt_forwarder {
+	LOG_RPC_EVT_MSG = 0,
+};
+
+enum log_rpc_cmd_backend {
+	LOG_RPC_CMD_GET_CRASH_LOG = 0,
+};
+
+#ifdef CONFIG_LOG_FORWARDER_RPC
+
+LOG_MODULE_REGISTER(rpc);
+
+static void log_rpc_msg_handler(const struct nrf_rpc_group *group, struct nrf_rpc_cbor_ctx *ctx,
+				void *handler_data)
+{
+	uint8_t level;
+	struct zcbor_string message;
+	bool decoded_ok;
+
+	decoded_ok = zcbor_uint_decode(ctx->zs, &level, sizeof(level));
+	decoded_ok = decoded_ok && zcbor_bstr_decode(ctx->zs, &message);
+	nrf_rpc_cbor_decoding_done(group, ctx);
+
+	if (!decoded_ok) {
+		goto out;
+	}
+
+	switch (level) {
+	case LOG_LEVEL_ERR:
+		LOG_ERR("%.*s", message.len, message.value);
+		break;
+	case LOG_LEVEL_WRN:
+		LOG_WRN("%.*s", message.len, message.value);
+		break;
+	case LOG_LEVEL_INF:
+		LOG_INF("%.*s", message.len, message.value);
+		break;
+	case LOG_LEVEL_DBG:
+		LOG_DBG("%.*s", message.len, message.value);
+		break;
+	default:
+		break;
+	}
+
+out:
+	return;
+}
+
+NRF_RPC_CBOR_EVT_DECODER(log_rpc_group, log_rpc_msg_handler, LOG_RPC_EVT_MSG, log_rpc_msg_handler,
+			 NULL);
+
+#endif /* CONFIG_LOG_FORWARDER_RPC */
+
+#ifdef CONFIG_LOG_BACKEND_RPC
+
+/*
+ * Drop messages coming from nRF RPC to avoid the avalanche of logs:
+ * 1. log added
+ * 2. log sent over nRF RPC
+ * 3. more logs generated by nRF RPC
+ * 4. more logs sent over nRF RPC
+ * ...
+ */
+static const char *const filtered_out_sources[] = {
+	"nrf_rpc",
+	"NRF_RPC",
+};
+
+static bool panic_mode;
+static uint32_t log_format = LOG_OUTPUT_TEXT;
+static uint8_t current_level;
+static uint8_t tx_buffer[CONFIG_LOG_BACKEND_RPC_BUFFER_SIZE];
+
+extern struct k_thread logging_thread;
+
+static int send(uint8_t *data, size_t length, void *context)
+{
+	struct nrf_rpc_cbor_ctx ctx;
+	const struct zcbor_string data_string = {
+		.value = data,
+		.len = length,
+	};
+
+	NRF_RPC_CBOR_ALLOC(&log_rpc_group, ctx, 4 + length);
```

Comment:
Just curious. Why `4 + length`?

Reply:
1 for log level (CBOR small uint)
3 for log string (1 for CBOR major, 2 for length)

But I agree, we need some sort of helper for that as current CBOR API is way too low level to be readable.